xml/System/Span`1+Enumerator.xml

```diff
@@ -21,8 +21,30 @@
     </Attribute>
   </Attributes>
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Provides an enumerator for the elements of a <see cref="T:System.Span`1" />.</summary>
+    <remarks>
+      <format type="text/markdown"><![CDATA[  
+
+The C# [foreach](~/docs/csharp/language-reference/keywords/foreach-in.md) of the C# language and the [For Each...Next](~/docs/visual-basic/language-reference/statements/for-each-next-statement.md) construct in Visual Basic hides the complexity of enumerators. Instead of directly manipulating the enumerator, using `foreach` or `For Each...Next` is recommended. 
+
+Enumerators can be used to read the data in the <xref:System.Span%601>, but they cannot be used to modify it. 
+
+Initially, the enumerator is positioned before the first element in the <xref:System.Span%601>. At this position, <xref:System.Span%601.Enumerator.Current> is undefined, so you must call <xref:System.Span%601.Enumerator.MoveNext%2A> to advance the enumerator to the first item in the <xref:System.Span%601> before reading the value of <xref:System.Span%601.Enumerator.Current>.
+
+<xref:System.Span%601.Enumerator.Current> returns the same object until <xref:System.Span%601.Enumerator.MoveNext%2A> is called. <xref:System.Span%601.Enumerator.MoveNext%2A> sets  <xref:System.Span%601.Enumerator.Current> to the next item in the <xref:System.Span%601>.
+
+If <xref:System.Span%601.Enumerator.MoveNext%2A> passes the end of the <xref:System.Span%601>, the enumerator is positioned after the last item in the <xref:System.Span%601>, and <xref:System.Span%601.Enumerator.MoveNext%2A> returns `false`. When the enumerator is at this position, subsequent calls to <xref:System.Span%601.Enumerator.MoveNext%2A> also return `false`. If the call to <xref:System.Span%601.Enumerator.MoveNext%2A> returns `false`, <xref:System.Span%601.Enumerator.Current> is undefined. You cannot set <xref:System.Span%601.Enumerator.Current> to the first item in the <xref:System.Span%601> again; you must create a new enumerator instance instead. 
+
+The enumerator does not have exclusive access to the <xref:System.Span%601>; therefore, enumerating through a span is intrinsically not a thread-safe procedure. To guarantee thread safety during enumeration, you can lock the span during the entire enumeration. To allow the span to be accessed by multiple threads for reading and writing, you must implement your own synchronization. 
```

Comment:
> To guarantee thread safety during enumeration, you can lock the span during the entire enumeration. 

@stephentoub, is this true? Couldn't the underlying array still be modified by another thread, or is the implication that the user needs some synchronization mechanism during enumeration and modification of the array? Also, how do we lock the span (it isn't a reference type).

The following snippet has a race. To resolve it, we have to use a locking mechanism around both critical sections.
**Output:**
62
23
0
0
0

```C#
static void Main(string[] args)
{
    new Random(42).NextBytes(_array);
    Span<byte> span = _array;

    Thread thread = new Thread(delegate ()
    {
        ClearContents();
    });

    thread.Start();

    EnumerateSpan(span);
}

private static readonly byte[] _array = new byte[5];

public static void ClearContents()
{
    Thread.Sleep(20);
    lock (_array)
    {
        Array.Clear(_array, 0, _array.Length);
    }
}

public static void EnumerateSpan(Span<byte> span)
{
    //lock (_array)
    //{
        foreach (byte element in span)
        {
            Console.WriteLine(element);
            Thread.Sleep(10);
        }
    //}
}
```

Reply:
First and foremost, there's no way to "lock the span".

Presumably this was really just meant to say that you can use your own synchronization to enforce mutual exclusion.